### PR TITLE
Update CRT libraries and set operation_name for DEFAULT meta-requests

### PR DESCRIPTION
**Pull request by passaro**
## Description of change

The latest `aws-c-s3` release ([v0.6.0](https://github.com/awslabs/aws-c-s3/releases/tag/v0.6.0)) introduces a breaking change: meta requests of type "Default" require users to set the `operation_name` field, which was previously optional. This change updates the CRT libraries to the latest version and ensures that `mountpoint-s3-client` always sets the operation name in all required cases.

<details>
<summary>Full CRT changelog</summary>
<pre>
Submodule mountpoint-s3-crt-sys/crt/aws-c-cal 96c47e3..11fc684:
  > Make AES GCM more consistent cross platform (#189)
  > Pin AWS-LC until it's fixed for manylinux1 (#188)
  > Implement runtime check on libcrypto linkage (#186)
  > clang-format 18 (#187)
Submodule mountpoint-s3-crt-sys/crt/aws-c-common 06cf4d8..6d974f9:
  > cbor support  (#1131)
  > Fix default thread options for windows to not pin to any cpu_id (#1126)
  > Use CBMC 6.0.0 (#1128)
  > latest_submodules.py uses AWS-LC-FIPS releases in aws-crt-java (#1125)
  > Use CBMC version 5.95.1 (#1124)
  > clang-format 18 (#1113)
  > disable optimization was not working (#1123)
  > Fix memtracer bad assumptions on the size of stack trace (#1122)
Submodule mountpoint-s3-crt-sys/crt/aws-c-s3 6588f9a..cb431ba:
  > test_helper.py improvements (#442)
  > Fix shutdown_callback or returning NULL contract for meta_request (#440)
  > BREAKING CHANGE: operation_name must be set for DEFAULT meta-requests (#439)
  > clang-format 18 (#438)
  > Auto - Update S3 Ruleset & Partition (#436)
Submodule mountpoint-s3-crt-sys/crt/aws-lc 92bf532..4368aaa:
  > Fix for loading JCA stripped private keys (#1658)
  > Prepare for release v1.30.1 (#1657)
  > Revert  `_CET_ENDBR` (#1656)
  > Close FD in Snapsafe test function (#1649)
  > Prepare for release v1.30.0 (#1646)
  > Snapsafe-type uniqueness breaking event detection (#1640)
  > Add EVP_md_null and SSL_set_ciphersuites (#1637)
  > Add de-randomized ML-KEM modes to experimental EVP API (#1578)
  > Patch for OpenVPN certificate setting behavioral difference (#1643)
  > Require newer assembler for _CET_ENDBR (#1641)
  > OpenVPN error codes, SSL_get_peer_signature_* funcs, and first patch file (#1584)
  > NIST.SP.800-56Cr2 One-Step Key Derivation (#1607)
  > Upstream merge 2024-06-13 (#1636)
  > More minor symbols for Ruby support (#1581)
  > Add support for NETSCAPE_SPKI_print (#1624)
  > align gcc version with curl's CI (#1633)
  > Fix spelling nits
  > Generated ASM files
  > Add Intel Indirect Branch Tracking support.
  > [EC] Unify point addition for P-256/384/521 (#1602)
  > Upstream merge 2024 06 03 (#1621)
  > Fix AES key size for AES256 in ABI test (#1629)
  > Move SSL_CIPHER_get_version test to SSLVersionTest.Version (#1631)
  > Use 'nasm' not 'yasm' (#1630)
  > Prepare for release 1.29.0 (#1626)
  > Implement SSL_CIPHER_get_version for recent TLS versions (#1627)
  > Add integration tests for OpenSSL-linking 3p modules (#1587)
  > Prevent non-constant-time code in Kyber-R3 and ML-KEM implementation (#1619)
  > Update ec2-test-framework to use gv2 (#1623)
  > Script for creating compilation database (#1617)
  > Fixes for building with `-pedantic` (#1608)
  > Fix SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR behavior (#1620)
  > Update for FIPS documentation (#1610)
  > Disable CI for gcc-14/FIPS until relocation issue is resolved (#1622)
  > Add support for ocsp get id (#1609)
  > Add libevent to GitHub integration CI (#1615)
  > Upstream merge 2024 05 17 (#1600)
  > add back ASN1_dup with tests (#1591)
  > Remove special aarch64 valgrind logic (#1618)
  > Fix NTP integ test (#1616)
  > Pin aws-lc-rs integ to nightly-2024-05-22 (#1612)
  > Cleanse the right amount of bytes in HMAC. (#1613)
  > add support for X509_CRL_http_nbio (#1596)
  > Add `all_fuzz_tests` build target (#1605)
  > Fix mariadb ssl_crl patch (#1606)
Submodule mountpoint-s3-crt-sys/crt/s2n-tls 6d92b46..073c7b4:
  > bug: Fixing bash error (#4624)
  > chore: make cbmc proof build more strict by adding -Werror flag (#4606)
  > Perform 2-RTT Handshake to upgrade to PQ when possible (#4526)
  > test(bindings/s2n-tls): refactor testing::s2n-tls tests (#4613)
  > docs: add timeout note to blinding delay docs (#4621)
  > docs: Add back suggested FIPS + TLS1.3 policy (#4605)
  > ci: shallow clone musl repo (#4611)
  > example(bindings): add async ConfigResolver (#4477)
  > chore: use CBMC version 5.95.1 (#4586)
  > s2n-tls rust binding: expose selected application protocol (#4599)
  > test: add pcap testing crate (#4604)
  > testing(bindings): add new test helper (#4596)
  > chore(bindings): fix shebang in generate.sh (#4603)
  > fix(s2n_session_ticket_test): correct clock mocking (#4602)
  > Fix: update default cert chain for unit tests (#4582)
  > refactor(binding): more accurate naming for const str helper (#4601)
  > fix: error rather than empty cipher suites (#4597)
  > chore: update s2n_stuffer_printf CBMC harness (#4531)
  > ci(nix): Fix integ pq test in a devShell (#4576)
  > feature: new compatibility-focused security policy preferring ECDSA (#4579)
  > compliance: update generate_report.sh to point to compliance directory (#4588)
  > ci: fix cppcheck errors (#4589)
  > chore: cleanup duplicate duvet citations (#4587)
  > Merge pull request from GHSA-52xf-5p2m-9wrv
  > chore(bindings): release 0.2.7 (#4580)
  > fix: Validate received signature algorithm in EVP verify (#4574)
  > refactor: add try_compile feature probe for RSA-PSS signing (#4569)
  > feat: Configurable blinding (#4562)
  > docs: document s2n_cert_auth_type behavior (#4454)
  > fix: init implicit iv for serialization feature (#4572)
  > [Nix] adjust pytest retrys (#4558)
  > fix: cert verify test fix (#4545)
  > fix: update default security policies (#4523)
  > feat(bindings): Associate an application context with a Connection (#4563)
  > chore(bindings): version bump (#4566)
  > Additional test cases for s2n_constant_time_equals() (#4559)
  > test: backwards compatibility test for the serialization feature (#4548)
  > chore(bench): upgrade rustls (#4554)
</pre>
</details>

## Does this change impact existing behavior?

No changes to the public API of `mountpoint-s3-client`.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3-client/src/s3_crt_client.rs [hunk 1]
```diff
@@ -744,6 +747,42 @@ impl S3CrtClientInner {
...
+            S3Operation::PutObject => MetaRequestType::PutObject,
+            _ => MetaRequestType::Default,
+        }
+    }
+
+    /// The operation name to set when configuring a request, if required.
+    fn operation_name(&self) -> Option<&'static str> {
+        match self {
+            S3Operation::DeleteObject => Some("DeleteObject"),
+            S3Operation::GetObject => None,
```
Comment (unexge) [inline mountpoint-s3-client/src/s3_crt_client.rs:776]: Is there any harm to always return operation name here even though it's not always required?

Comment (passaro) [reply mountpoint-s3-client/src/s3_crt_client.rs:776]: The CRT docs only say it's required for Default meta requests. It's likely ignored in other cases, but it seems sensible to leave it unset: e.g. which operation would we use for a "PutObject" meta-request, which may translate into CreateMPU, UploadPart, etc.?

Comment (arsh) [reply mountpoint-s3-client/src/s3_crt_client.rs:776]: >which operation would we use for a "PutObject" meta-request

This is transparent to us, the users of CRT. So, it should not matter. If CRT decides to split an operation, it is their business.

I don't think this is a blocking thing, but I can see how it is more simple to always return and set an operation name.

File: mountpoint-s3-client/src/s3_crt_client.rs [hunk 2]
```diff
@@ -744,6 +747,42 @@ impl S3CrtClientInner {
...
+    DeleteObject,
+    GetObject,
+    GetObjectAttributes,
+    HeadBucket,
+    HeadObject,
+    ListObjects,
+    PutObject,
+}
+
+impl S3Operation {
```
Comment (arsh) [inline mountpoint-s3-client/src/s3_crt_client.rs:762]: why aren't there unit tests verifying that we return the right operation names and request types?

Comment (passaro) [reply mountpoint-s3-client/src/s3_crt_client.rs:762]: I'm not convinced unit tests of these 2 methods would add any value. On the other hand, we do have integration tests that will use the results to initiate S3 requests.

File: mountpoint-s3-client/src/s3_crt_client.rs (hunk 1 above)
Comment (arsh) [reply mountpoint-s3-client/src/s3_crt_client.rs:776]: (Related to the unit test question) How do you stick to their guidance without a test that fails if we anyone intends to send an operation name for GET or PUT?

**Review by arsh — APPROVED** [review]

File: mountpoint-s3-client/src/s3_crt_client.rs (hunk 1 above)
Comment (passaro) [reply mountpoint-s3-client/src/s3_crt_client.rs:776]: @arsh, maybe I am missing something. What kind of test do you have in mind?

Comment (arsh) [reply mountpoint-s3-client/src/s3_crt_client.rs:776]: For example, if you think we should not be sending an operation name for PUT and GET, then a unit test can help enforce that. With those tests, you are codifying your current assumption until those change. Without any tests, I can start returning an operation name for those any time and no test will fail.

File: mountpoint-s3-client/src/s3_crt_client.rs (hunk 2 above)
Comment (vladem) [reply mountpoint-s3-client/src/s3_crt_client.rs:762]: also thought about a unit test that verifies that for each variant of `S3Operation`, which results in `MetaRequestType::Default`, there is a not-empty `operation_name`  

but can't think of a way of writing such a test in a good way, and in the worst case (if we ever decide to use another s3 operation with default meta-request, but with no `operation_name`) we'll receive descriptive (hopefully) error from crt on request creation and catch it in a development phase

**Pull request merged**